New config.yml option: `afk-timeout-command:`

ZepsiZola opened the pull request:
Implements this https://github.com/EssentialsX/Essentials/issues/2864

**Specifically, this functionality:**

`afk-timeout-command: "server afk {USERNAME}"`
Sends the player to the AFK server on the server-network.

`afk-timeout-command: "litebans:kick {USERNAME} You were kicked after {KICK_TIME} minutes of inactivity."`
Kicks the player using litebans so that it shows up in their litebans history.

`afk-timeout-command: "none"`
Default behaviour of essentials kicking the player on afk-timeout.

ZepsiZola committed:
Added config option

ZepsiZola committed:
Reversed auto-formatted code

ZepsiZola committed:
Updated .gitignore to ignore  and updated config.yml

ZepsiZola committed:
Updated .gitignore to ignore bin/ directories

JasonHorkles commented on Essentials/src/main/resources/config.yml:
```diff
@@ -484,6 +484,11 @@ auto-afk: 300
 
+# Define the command that gets run by the server when the player's AFK time elapses.
+# Keep as 'none' to use Essentials default afk-timeout kicking behaviour.
+# Available placeholders: {PLAYER}, {KICK_TIME}
```
```suggestion
# Define a command the server runs when a player's AFK time elapses.
# Set to 'none' to use Essentials' default AFK timeout kick behavior.
#
# Available placeholders:
# {USERNAME} - The player's username.
# {KICKTIME} - The time, in minutes, the player has been AFK for.
```

JasonHorkles commented on Essentials/src/main/java/com/earth2me/essentials/User.java:
```diff
@@ -825,20 +825,30 @@ public void checkActivity() {
             }
+            // If `afk-timeout-command` in config.yml is not set to 'none', execute the command instead of kicking the player.
+            else {
+                final String command = ess.getSettings().getAfkTimeoutCommand().replace("{PLAYER}", getName()).replace("{KICK_TIME}", String.valueOf(kickTime));
```
```suggestion
                final String command = ess.getSettings().getAfkTimeoutCommand().replace("{USERNAME}", getName()).replace("{KICKTIME}", String.valueOf(kickTime));
```

ZepsiZola committed:
Update Essentials/src/main/java/com/earth2me/essentials/User.java

Co-authored-by: Jason <51760372+JasonHorkles@users.noreply.github.com>

ZepsiZola committed:
Update Essentials/src/main/resources/config.yml

Co-authored-by: Jason <51760372+JasonHorkles@users.noreply.github.com>

JRoy commented on .gitignore:
```diff
@@ -25,6 +25,7 @@ out/
 build/
 target/
 *.class
+*/bin/
```
no reason to add this

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/Settings.java:
```diff
@@ -1266,6 +1266,11 @@ public long getAutoAfkKick() {
 
+    @Override
+    public String getAfkTimeoutCommand() {
+        return FormatUtil.replaceFormat(config.getString("afk-timeout-command", "none"));
```
why would we replaceFormat here? you cannot send section signs thru commands

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/User.java:
```diff
@@ -825,20 +825,30 @@ public void checkActivity() {
+                        user.sendTl("playerKicked", Console.DISPLAY_NAME, getName(), user.playerTl("autoAfkKickReason", kickTime));
+                    }
                 }
             }
+            // If `afk-timeout-command` in config.yml is not set to 'none', execute the command instead of kicking the player.
+            else {
```
```suggestion
            } else {
                // If `afk-timeout-command` in config.yml is not set to 'none', execute the command instead of kicking the player.
```

JRoy commented on Essentials/src/main/resources/config.yml:
```diff
@@ -484,6 +484,14 @@ auto-afk: 300
+#
+# Available placeholders:
+# {USERNAME} - The player's username.
+# {KICKTIME} - The time, in minutes, the player has been AFK for.
```
is there a reason we need this variable, wouldn't the time the player has been afk always be the same (the auto-afk-kick above).

JRoy commented on Essentials/src/main/resources/config.yml:
```diff
@@ -484,6 +484,14 @@ auto-afk: 300
 # Set to -1 for no timeout.
 auto-afk-kick: -1
```
maybe we rename this option to `auto-afk-timeout` and explain that it will run the command below if set, otherwise kick the player?

ZepsiZola replied:
> no reason to add this

I added it because this is what I got after doing `./gradlew` and then `./gradlew shadowJar`. This was directly after cloning the repo for the first time.
![unknown](https://github.com/user-attachments/assets/7630de21-448c-457b-911e-07aa4c2e616a)
You can reject this change if you want, but I added it because there were unneeded files created on my repo that I did not want to push.

> why would we replaceFormat here? you cannot send section signs thru commands

Fair enough. I wasn't familiar with the EssentialsX codebase and thought that that was needed if I wanted to parse placeholders in the config option. I can change that line to not use FormatUtil.


> is there a reason we need this variable, wouldn't the time the player has been afk always be the same (the auto-afk-kick above).

The variable is there so that if a server owner changes the AFK timeout, they don't also have to change the `afk-timeout-command`. And no, the `auto-afk-kick` doesn't always stay the same. As a server owner myself I've changed it around a fair few times. It would be an inconvenience to anyone messing with their afk configuration to have to change a number in two places because they changed the `auto-afk-kick` config. A minor inconvenience, but a completely unneeded one as well.

> maybe we rename this option to `auto-afk-timeout` and explain that it will run the command below if set, otherwise kick the player?

@JRoy `auto-afk-timeout` is bad naming. From the name, it would seem like it should define the timeout for a player's AFK. I don't think it should be named that *just* because the other config options have "auto-afk" in them. We shouldn't give something a bad name on purpose just so that it looks similar to the other config options. It should be called `afk-timeout-command` because it is the command that is run upon the players AFK-time running out. Isn't the fact that the config option is right below `auto-afk-kick` enough for it to be understood that it closely relates to that config option?

JRoy replied:
still want this renamed + docs updated to explain that this will only kick people if the below option isn't set

JRoy reviewed (changes requested):
have a few issues with this, would also like @mdcfe's input on this

JRoy reviewed (changes requested):
have a few issues with this, would also like @mdcfe's input on this

ZepsiZola commented:
> no reason to add this

I added it because this is what I got after doing `./gradlew` and then `./gradlew shadowJar`. This was directly after cloning the repo for the first time.
![unknown](https://github.com/user-attachments/assets/7630de21-448c-457b-911e-07aa4c2e616a)
You can reject this change if you want, but I added it because there were unneeded files created on my repo that I did not want to push.

> why would we replaceFormat here? you cannot send section signs thru commands

Fair enough. I wasn't familiar with the EssentialsX codebase and thought that that was needed if I wanted to parse placeholders in the config option. I can change that line to not use FormatUtil.


> is there a reason we need this variable, wouldn't the time the player has been afk always be the same (the auto-afk-kick above).

The variable is there so that if a server owner changes the AFK timeout, they don't also have to change the `afk-timeout-command`. And no, the `auto-afk-kick` doesn't always stay the same. As a server owner myself I've changed it around a fair few times. It would be an inconvenience to anyone messing with their afk configuration to have to change a number in two places because they changed the `auto-afk-kick` config. A minor inconvenience, but a completely unneeded one as well.

> maybe we rename this option to `auto-afk-timeout` and explain that it will run the command below if set, otherwise kick the player?

@JRoy `auto-afk-timeout` is bad naming. From the name, it would seem like it should define the timeout for a player's AFK. I don't think it should be named that *just* because the other config options have "auto-afk" in them. We shouldn't give something a bad name on purpose just so that it looks similar to the other config options. It should be called `afk-timeout-command` because it is the command that is run upon the players AFK-time running out. Isn't the fact that the config option is right below `auto-afk-kick` enough for it to be understood that it closely relates to that config option?

SrBedrock commented:
I believe this could be a list of commands and instead of none, check if the list is empty, just like other settings already do

ZepsiZola committed:
Update Essentials/src/main/java/com/earth2me/essentials/User.java

Co-authored-by: Josh Roy <10731363+JRoy@users.noreply.github.com>

JasonHorkles commented:
> I added it because there were unneeded files created on my repo that I did not want to push.

You can deselect items that you don't want to commit 😉 

> It should be called `afk-timeout-command` because it is the command that is run upon the players AFK-time running out.

Josh was referring to the older config option `auto-afk-kick` (not the one you're adding)

---

> I believe this could be a list of commands and instead of none, check if the list is empty, just like other settings already do

This is something that the team seems to agree with as well, although it's still being debated

ZepsiZola commented:
> > I added it because there were unneeded files created on my repo that I did not want to push.
> 
> You can deselect items that you don't want to commit 😉

I know that. But isn't the point of .gitignore to ignore files that will never be comitted? Unless this repo does require files in bin/ directories being pushed.

> > It should be called `afk-timeout-command` because it is the command that is run upon the players AFK-time running out.
> 
> Josh was referring to the older config option `auto-afk-kick` (not the one you're adding)

Ahh he was suggesting changing the existing config option name. Fair enough. I'm impartial to that.

> > I believe this could be a list of commands and instead of none, check if the list is empty, just like other settings already do
> 
> This is something that the team seems to agree with as well, although it's still being debated

I can implement this if you want. I don't necessarily care about that functionality, but it won't interfere with my own functionality, plus I just want this feature implemented so I can use it in my server as soon as I can.

ZepsiZola committed:
Changes

Renamed config option to afk-timeout-commands and made it accept a list of strings.

Removed unnecessary FormatUtil.

Reverted .gitignore

ZepsiZola commented:
@SrBedrock @JasonHorkles @JRoy 
Btw I made these changes:
- Reverted the .gitignore change. So it remains unchanged from before the PR.
- Removed use of FormatUtil where not needed.
- Made it so you can specify a list of commands that run in succession as opposed to just one command. Config option is now called `afk-timeout-commands:`

Now the config.yml looks like this...
```YAML
# Define a set of commands the server runs when a player's AFK time elapses.
# Set to [] to use Essentials' default AFK timeout kick behavior.
#
# Available placeholders:
# {USERNAME} - The player's username.
# {KICKTIME} - The time, in minutes, the player has been AFK for.
afk-timeout-commands: []
```
and you can set commands like this
```YAML
afk-timeout-commands:
- "msg {USERNAME} You have been AFK for {KICKTIME}. Jailing you and moving you to AFK server..."
- "jail {USERNAME} jail-1"
- "server afk {USERNAME}"
```

JasonHorkles commented:
Seems fine to me, although I do have a concern about your possible intended use for this - the /server command is created by the proxy and not the backend servers, so unless you have a custom server switching solution, the /server command won't work

ZepsiZola commented:
> Seems fine to me, although I do have a concern about your possible intended use for this - the /server command is created by the proxy and not the backend servers, so unless you have a custom server switching solution, the /server command won't work

I have a plugin that creates `/server` commands on the backend servers. Don't worry.

ZepsiZola commented:
Do I need to "Mark as resolved" on requested changes or can I just leave this?

JasonHorkles commented:
Can't hurt to resolve ones you implemented

ZepsiZola commented:
> Can't hurt to resolve ones you implemented

I've done that, but what about the ones I haven't implemented?

https://github.com/EssentialsX/Essentials/pull/6169#discussion_r2144125450
https://github.com/EssentialsX/Essentials/pull/6169#discussion_r2144125450

JasonHorkles commented:
Leave them open for now

JRoy committed:
Merge branch '2.x' into afk-timeout-command

JRoy commented:
> ./gradlew and then ./gradlew shadowJar

run `./gradlew build` instead, jars will appear in the `jars` folder in the root directory. this shouldn't create any bin folders.

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/Settings.java:
```diff
@@ -1266,6 +1268,17 @@ public long getAutoAfkKick() {
+    @Override
+    public List<String> getAfkTimeoutCommands() {
+        return afkTimeoutCommands;
+    }
```
```suggestion
    private Set<String> _getAfkTimeoutCommands() {
        final Set<String> timeoutCommands = new HashSet<>();
        for (final String cmd : config.getList("afk-timeout-commands", String.class)) {
            timeoutCommands.add(cmd.toLowerCase(Locale.ENGLISH));
        }
        return timeoutCommands;
    }

    @Override
    public Set<String> getAfkTimeoutCommands() {
        return afkTimeoutCommands;
    }
```

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/Settings.java:
```diff
@@ -153,6 +153,7 @@ public class Settings implements net.ess3.api.ISettings {
     private Tag secondaryColor = DEFAULT_SECONDARY_COLOR;
     private Set<String> multiplierPerms;
     private BigDecimal defaultMultiplier;
+    private List<String> afkTimeoutCommands = Collections.emptyList();
```
```suggestion
    private Set<String> afkTimeoutCommands = new HashSet<>();
```

JRoy commented on Essentials/src/main/resources/config.yml:
```diff
@@ -484,6 +484,14 @@ auto-afk: 300
 # Set to -1 for no timeout.
 auto-afk-kick: -1
 
+# Define a set of commands the server runs when a player's AFK time elapses.
+# Set to [] to use Essentials' default AFK timeout kick behavior.
+#
+# Available placeholders:
+# {USERNAME} - The player's username.
+# {KICKTIME} - The time, in minutes, the player has been AFK for.
+afk-timeout-commands: []
```
```suggestion
# A list of commands to be executed instead of kicking the player once the
# threshold defined above in 'afk-auto-timeout' is reached. If this list is empty 
# and 'afk-auto-timeout' is not set to -1, Essentials will default to
# kicking the player once they reach the timeout threshold.
#
# WARNING: You must include a command here that either removes the player from the server
# or stops them from being AFK. Otherwise, these commands will run every second
# until the player is no longer AFK!
#
# Available placeholders:
# {USERNAME} - The player's username.
# {KICKTIME} - The time, in minutes, the player has been AFK for.
afk-timeout-commands:
  #- eco take {USERNAME} 10
  #- kick {USERNAME} You have been kicked for being inactive for {KICKTIME} minutes! You lost $10.
```

JRoy reviewed (changes requested):
few things

ZepsiZola committed:
Update Essentials/src/main/resources/config.yml

Co-authored-by: Josh Roy <10731363+JRoy@users.noreply.github.com>

ZepsiZola committed:
Update Essentials/src/main/java/com/earth2me/essentials/Settings.java

Co-authored-by: Josh Roy <10731363+JRoy@users.noreply.github.com>

ZepsiZola committed:
Update Essentials/src/main/java/com/earth2me/essentials/Settings.java

Co-authored-by: Josh Roy <10731363+JRoy@users.noreply.github.com>

ZepsiZola commented:
https://github.com/EssentialsX/Essentials/pull/6169#discussion_r2160418211

Can I just rename the config option? Wouldn't that affect people with older configs? Or is there more I have to do when renaming a config option?
I'd rename it to `auto-afk-timeout`.

Also when you say to update the docs, which docs do you mean and can I even update them within this repo? I thought the docs were hosted on https://essentialsx.net/wiki

JRoy commented:
> Can I just rename the config option? Wouldn't that affect people with older configs? Or is there more I have to do when renaming a config option?

Look at this example to see how to do it, you would look for the new name, if that isn't set look for the old name, if that isn't set, use the default value https://github.com/EssentialsX/Essentials/blob/ff1b8b8615d72cdd8b34a37cfd0c2766b83ad86a/Essentials/src/main/java/com/earth2me/essentials/Settings.java#L222-L224

> Also when you say to update the docs, which docs do you mean

Just the comments above the option in the config.yml

ZepsiZola committed:
Changed config option from auto-afk-kick to auto-afk-timeout

ZepsiZola committed:
Changed config option from auto-afk-kick to auto-afk-timeout

ZepsiZola committed:
Changed config option from auto-afk-kick to auto-afk-timeout

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/User.java:
```diff
@@ -832,19 +832,34 @@ public void checkActivity() {
+                    }
+                }
+            } else {
+                // If `afk-timeout-command` in config.yml is populated, execute the command(s) instead of kicking the player.
```
```suggestion
                // If `afk-timeout-commands` in config.yml is populated, execute the command(s) instead of kicking the player.
```

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/User.java:
```diff
@@ -832,19 +832,34 @@ public void checkActivity() {
-                && lastActivity > 0 && (lastActivity + (autoafkkick * 1000)) < System.currentTimeMillis()
+        final long autoafktimeout = ess.getSettings().getAutoAfkTimeout();
+
+        // Checks if the player has been inactive for longer than the configured auto-afk-kick time.
```
```suggestion
        // Checks if the player has been inactive for longer than the configured auto-afk-timeout time.
```

JRoy commented on Essentials/src/main/resources/config.yml:
```diff
@@ -484,6 +484,22 @@ auto-afk: 300
 # Set to -1 for no timeout.
 auto-afk-kick: -1
```
```suggestion
auto-afk-timeout: -1
```

JRoy committed:
Apply suggestions from code review

JRoy commented on Essentials/src/main/resources/config.yml:
```diff
@@ -482,7 +482,23 @@ auto-afk: 300
 # After this timeout in seconds, the player will be kicked from the server.
 # The 'essentials.afk.kickexempt' permission overrides this feature.
 # Set to -1 for no timeout.
```
```suggestion
# After this timeout in seconds, the player will either be kicked from
# the server or commands in 'afk-timeout-commands' will be executed.
# The 'essentials.afk.kickexempt' permission overrides this feature.
# Set to -1 for no timeout.
```

JRoy committed:
Update Essentials/src/main/resources/config.yml

JRoy reviewed (approved):
lgtm, will merge after our next release

forsakenliquid commented:
when will be the rollout? (release)

JasonHorkles commented:
> when will be the rollout? (release)

It'll be released when it gets released. We don't have any ETAs

ZepsiZola commented:
Hell yeah.
